src/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java

```diff
@@ -0,0 +1,187 @@
+/**
+ * The XSLTInjection plugin identifies XSLT injection vulnerabilities with all the parameters in GET and POST
+ *
+ * @author CaptainFreak
+ */
+public class XSLTInjection extends AbstractAppParamPlugin {
+
+	// wasc_23 is XML Injection
+	private static Vulnerability vuln = Vulnerabilities.getVulnerability("wasc_23");
+    
+	// plugin dependencies
+    private static final String[] dependency = {};
+    
+	//For logging
+	private static Logger log = Logger.getLogger(XSLTInjection.class);
+    private boolean debugEnabled = log.isDebugEnabled();
+
+    // Payloads
+    
+    private static final String[] errorCausingPayloads={"<",">","\'","\""};
+    
+    private static final String[] vendorReturningPayloads={"<xsl:value-of select=\"system-property(\'xsl:vendor\')\"/>",
+    														"system-property(\'xsl:vendor\')/>",
+    														"\"/><xsl:value-of select=\"system-property(\'xsl:vendor\')\"/><!--",
+    														"<xsl:value-of select=\"system-property(\'xsl:vendor\')\"/><!--"
+    														};
+    private static final String[] xsltVendors={"libxslt","Microsoft","Saxonica","Apache","Xalan","SAXON","Transformiix"};
+    
+    //TODO Add time based payloads (SSRF)
+    
+	@Override
+	public int getId() {
+		return 90017;
+	}
+
+	@Override
+	public String getName() {
+		
+		return Constant.messages.getString("ascanalpha.xsltinjection.xslt.name");
+	}
+
+	@Override
+	public String[] getDependency() {
+		return dependency;
+	}
+
+	@Override
+	public boolean targets(TechSet technologies) {  
+		return technologies.includes(Tech.XML); 
+	}
+	
+	@Override
+	public String getDescription() {
+		return Constant.messages.getString("ascanalpha.xsltinjection.xslt.desc");
+	}
+
+	@Override
+	public int getCategory() {
+		return Category.INJECTION;
+	}
+
+	@Override
+	public String getSolution() {
+		return Constant.messages.getString("ascanalpha.xsltinjection.xslt.soln");
+	}
+
+	@Override
+	public String getReference() {
+		return Constant.messages.getString("ascanalpha.xsltinjection.xslt.refs");
+	}
+
+	@Override
+	public void init() {
+        //DEBUG: turn on for debugging
+        //log.setLevel(org.apache.log4j.Level.DEBUG);
+        //this.debugEnabled = true;
+
+        if (this.debugEnabled) {
+            log.debug("Initialising XSLT Scanner.");
+        }
+
+	}
+
+	/*
+	 * This method is called by the active scanner for each GET and POST parameter for every page 
+	 * @see org.parosproxy.paros.core.scanner.AbstractAppParamPlugin#scan(org.parosproxy.paros.network.HttpMessage, java.lang.String, java.lang.String)
+	 */
+	
+	@Override
+	public void scan(HttpMessage msg, String param, String value) {
+		try {
+			//Count positive Tests
+			int positiveBasedOnError=0; 
+			//error causing tests
+			for (int i=0;i<errorCausingPayloads.length;i++){
+				msg = getNewMsg();
+				setParameter(msg, param, errorCausingPayloads[i]);
+				sendAndReceive(msg);
+				if(!msg.getResponseBody().toString().contains(errorCausingPayloads[i])){
```

Comment:
Isn't this logic backward? Shouldn't the counter only be incremented if the payload is found?

Reply:
No. As in this case payloads are special characters of xslt and injecting them cause compile time errors and in realistic scenario (PHP7 + libxslt) like the one I used, It doesn't return error or anything. It just drops the whole element from DOM (also production apps never return errors). This tests will be improved over time after researching in various xslt processors and real life scenarios but its really obscure vuln and hence all the accuracy testing will be done on self written vuln sites. I will do it over the time.

Reply:
Ok

Reply:
> production apps never return errors

I agree with properly configured apps, that's not always the case. Be good to check for errors as well.